### PR TITLE
Sync with upstream

### PR DESCRIPTION
**Pull request by jacob-hughes**
The directory tree in rustc has completely changed so there were a lot of merge conflicts to fix here. Note that most of our changes now reside in `library/alloc/src/` instead of `src/liballoc/`.  There have been a few breaking API changes too, and these have been addressed in the commits on the tip of tree.

### CODE REVIEW

**Comment by jacob-hughes** [comment]
Ah. The buildbot script was relying on the old `src/libstd` directory. I've updated it so it should work now.

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
Ah, rustfmt! This will need squashing once you're happy with it.

bors try

**Comment by ltratt** [comment]
I think you need another go :)

**Comment by jacob-hughes** [comment]
bors try

**Comment by jacob-hughes** [comment]
This is ready for re-review now

File: library/boehm_shim/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "boehm_shim"
+version = "0.1.0"
+authors = ["Jacob Hughes <jh@jakehughes.uk>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (ltratt) [inline library/boehm_shim/Cargo.toml:7]: Not sure we need this line?

Comment (jacob-hughes) [reply library/boehm_shim/Cargo.toml:7]: Good catch. Copy / paste autopilot

Comment (jacob-hughes) [reply library/boehm_shim/Cargo.toml]: Fixed: f055f76525faed922ad03743d6c337287e22bea7

**Comment by jacob-hughes** [comment]
Thank god!

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**